File: problems/1123-lowest-common-ancestor-of-deepest-leaves/solution.cpp
// Link: https://leetcode.cn/problems/lowest-common-ancestor-of-deepest-leaves/
// Level: Medium
// Date: 2025-04-04

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    public:
        pair<TreeNode*, int> f(TreeNode* root) {
            if (!root)
                return {nullptr, 0};
            auto [left, l] = f(root->left);
            auto [right, r] = f(root->right);
            if (l == r)
                return {root, l + 1};
            return l > r ? pair{left, l + 1} : pair{right, r + 1};
        }

        TreeNode* lcaDeepestLeaves(TreeNode* root) {
            return f(root).first;
        }
};

File: problems/1123-lowest-common-ancestor-of-deepest-leaves/solution.py
# Link: https://leetcode.cn/problems/lowest-common-ancestor-of-deepest-leaves/
# Level: Medium
# Date: 2025-04-04

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def f(root):
            if not root:
                return 0, None
            
            d1, lca1 = f(root.left)
            d2, lca2 = f(root.right)

            if d1 > d2:
                return d1 + 1, lca1
            if d1 < d2:
                return d2 + 1, lca2
            return d1 + 1, root

        return f(root)[1]